Add argument type information

louyihua opened the pull request:
The parse_arguments function does not maintain argument type information.
Here a new function, parse_arguments_with_type is provided, so that each arg's type info is stored and users can easily determine whether a argument is a string literal or not.

louyihua commented:
@sarina @nedbat
This is my patch to markey, which keeps the argument's type information. By using the newly added 'parse_arguments_with_type' function, users can easily distinguish string literals from varible names or other things, so in the django_babel_underscore, the logic will be much more simplifid.

sarina commented:
Related to bugs uncovered in https://github.com/edx/edx-platform/pull/10416

sarina commented on src/markey/machine.py:
```diff
@@ -113,6 +114,23 @@ def parse_arguments(stream, end_token):
+    args, kwargs = parse_arguments_with_type(stream, end_token)
+
+    raw_args = []
+    for arg, type in args:
```
This isn't very performant, and is the wrong way of doing code reuse. I think the better route is to modify the original implementation (L89-110) in a helper method, that takes a default-false arg `extract_types`. So then you have a function like this:

``` python
def parse_arguments_helper(stream, end_token, extract_types=False):
    # code here
    # L104 code:
    if extract_types:
        args.append((value, 'func_string_arg'))
    else:
        args.append(value)
    # continue rest of code, flipping what is appended to args based on extract_types value
```

Now, define your two main functions very cleanly:

``` python
def parse_arguments(stream, end_token):
    return parse_arguments_helper(stream, end_token)


def parse_arguments_with_type(stream, end_token):
    return parse_arguments_helper(stream, end_token, extract_types=True)
```

sarina commented:
@EnTeQuAk it looks like this build is failing because of how tests are defined:

```
2.6 runtests: commands[2] | py.test tests/
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: --clearcache
  inifile: /home/travis/build/EnTeQuAk/markey/pytest.ini
  rootdir: /home/travis/build/EnTeQuAk/markey
ERROR: InvocationError: '/home/travis/build/EnTeQuAk/markey/.tox/2.6/bin/py.test tests/'
___________________________________ summary ____________________________________
ERROR:   2.6: commands failed
```

EnTeQuAk commented:
Thanks a lot for the pull requests! I think this makes sense, one idea though, how about make the `type` extraction default and avoid having two methods for that?

Also, I'd like to re-use `markey.tools.Token`, if we're just appending the current token in the stream / create a token for `kwargs` we'd have extendable objects that probably would make usage easier.

Any ideas on that?

louyihua commented:
@sarina @EnTeQuAk 
To not break any existing code that may use markey while providing ability to give type information, I think it is enough to just adding a new parameter in the `parse_arguments` function. The new `extract_type_info` argument is set to `False` in default for backward compatibility, while setting `True` for this argument will make the function returns type information in `args`.

EnTeQuAk commented:
I'm not really a fan of returning differently typed results based on an argument. I kindof doubt that markey is widely used and besides that we can always create a major release and break compatibility.

louyihua commented:
@EnTeQuAk
Since no compatibility should be kept, just changing the function to return type info is enough.

sarina commented:
@EnTeQuAk Fair point. @louyihua if the library maintainer is OK w/ breaking compatibility, let's go for it!

Commit:
Add argument type information

The parse_arguments function does not maintain argument type information.
Here a new function, parse_arguments_with_type is provided, so that each arg's type info is stored and users can easily determine whether a argument is a string literal or not.

EnTeQuAk commented:
Awesome, thanks a lot! :cake: :beers:

louyihua commented:
@EnTeQuAk
Maybe this version can be labeled as 0.8?
PR#4 in `django_babel_underscore` (https://github.com/EnTeQuAk/django-babel-underscore/pull/4) requires this new version to work properly.

EnTeQuAk commented:
I will probably do that, I'd like to go over docs and unittests this weekend though so I might release this some time next week.

sarina commented:
Great news re the new release. We're finishing our upgrade to Django 1.8 now, and once that is complete will want to pull in the latest versions of marky and django-babel-underscore.
